include/jemalloc/internal/sized_alloc_region.h

```diff
@@ -0,0 +1,144 @@
+#ifndef JEMALLOC_INTERNAL_SIZED_ALLOC_REGION_H
+#define JEMALLOC_INTERNAL_SIZED_ALLOC_REGION_H
+
+#include "jemalloc/internal/alloc_ctx.h"
+#include "jemalloc/internal/assert.h"
+#include "jemalloc/internal/atomic.h"
+#include "jemalloc/internal/pages.h"
+
+/*
+ * This file defines sized alloc regions.  These regions have portions of memory
+ * dedicated to a particular size class.  When allocations come out of these
+ * regions, we can quickly determine their size based only on their address and
+ * a little bit of global metadata, thus avoiding any rtree lookups or (if we're
+ * lucky) any other cache-unfriendly lookups at all.
+ *
+ * We orchestrate it so that zero-initialization is a valid state for the
+ * sized_alloc_region_t.  This lets us avoid branching on
+ * opt_sized_alloc_region.
+ */
+
+/*
+ * SIZED_ALLOC_REGION_LG_NUM_SCS:
+ *     The log of the number of size classes that can come out of the
+ *     sized-alloc region.
+ * SIZED_ALLOC_REGION_LG_SC_SIZE:
+ *     The log of the size of virtual address space that we allocate to each
+ *     size class within the sized-alloc region.
+ */
+#if LG_SIZEOF_PTR == 2
```

Comment:
Oh, and our application memory usage is probably a bit atypical. We cap the VMEM of the process because if we have a memory leak we'd rather crash, report on it, and restart than page. We can't afford to page. Memory is often allocated by one CPU and free'd on another. We probably have windows of time where one CPU does lots of allocations and no frees, and another CPU is doing lots of frees and no allocations.

I experimented with turning off tcache and I did measure a noticeable drop in performance, so we're definitely getting something from that.

We would prefer everything to be lock free since we start up one thread per CPU core and then have an internal task system. Anytime a thread takes a lock (or enters the kernel which often means a lock) we can end up both starving the system of work (multiple threads are blocked on a lock and not running our tasks), and it also can cause a latency spike for the operation that was unfortunate enough to need to wait on a lock or call a syscall.